core/src/main/java/com/linecorp/armeria/server/RouteDecoratingService.java

```diff
@@ -0,0 +1,144 @@
+ *   <li>Finds all matched decorators from {@link InitialDispatcherService#router}
+ *       using {@link RoutingContext}</li>
+ *   <li>Put them into {@code pendingDecorators} with keeping the finding order</li>
+ *   <li>Finally, put the original {@link Service} at the end of {@code pendingDecorators}</li>
+ * </ul>
+ *
+ * <p>The request will go through the below decorators to reach the {@code userService}.
+ * <pre>{@code
+ *  request -> initialDispatcherService
+ *          -> loggingDecorator         -> routeDecoratingService
+ *          -> authDecorator            -> routeDecoratingService
+ *          -> traceDecorator           -> routeDecoratingService
+ *          -> userService
+ * }</pre>
+ */
+final class RouteDecoratingService implements HttpService {
+
+    private static final AttributeKey<Queue<Service<HttpRequest, HttpResponse>>> DECORATOR_KEY =
+            AttributeKey.valueOf(RouteDecoratingService.class, "PENDING_DECORATORS");
+
+    static Function<Service<HttpRequest, HttpResponse>,
+            Service<HttpRequest, HttpResponse>> newDecorator(Router<RouteDecoratingService> router) {
+        return delegate -> new InitialDispatcherService(delegate, router);
+    }
+
+    private final Route route;
+    private final Service<HttpRequest, HttpResponse> decorator;
+
+    RouteDecoratingService(Route route,
+                           Function<Service<HttpRequest, HttpResponse>,
+                                   ? extends Service<HttpRequest, HttpResponse>> decoratorFunction) {
+        this.route = requireNonNull(route, "route");
+        decorator = requireNonNull(decoratorFunction, "decoratorFunction").apply(this);
+    }
+
+    @Override
+    public HttpResponse serve(ServiceRequestContext ctx, HttpRequest req) throws Exception {
+        final Queue<Service<HttpRequest, HttpResponse>> delegates = ctx.attr(DECORATOR_KEY).get();
+        final Service<HttpRequest, HttpResponse> delegate = delegates.poll();
+        return delegate.serve(ctx, req);
+    }
+
+    Route route() {
+        return route;
+    }
+
+    private Service<HttpRequest, HttpResponse> decorator() {
+        return decorator;
+    }
+
+    @Override
+    public String toString() {
+        return MoreObjects.toStringHelper(this)
+                          .add("route", route)
+                          .add("decorator", decorator).toString();
+    }
+
+    private static class InitialDispatcherService extends SimpleDecoratingHttpService {
+
+        private final Router<RouteDecoratingService> router;
+
+        InitialDispatcherService(Service<HttpRequest, HttpResponse> delegate,
+                                 Router<RouteDecoratingService> router) {
+            super(delegate);
+            this.router = router;
+        }
+
+        @Override
+        public HttpResponse serve(ServiceRequestContext ctx, HttpRequest req) throws Exception {
+            final Queue<Service<HttpRequest, HttpResponse>> pendingDecorators = new ArrayDeque<>();
```

Comment:
Maybe better to specify a capacity than the default 16 (4?) - 16 decorators seems quite improbable.

Reply:
Might call this `serviceChain` - the last element in it isn't a decorator.

Reply:
> Might call this serviceChain - the last element in it isn't a decorator.

IIUC, you meant renaming `decorator.serve(ctx, req)` -> `service.serve(ctx, req)`?

Reply:
Nah just meant renaming `pendingDecorators` to `serviceChain` since we `.add(delegate())` at the end

Reply:
I see, Thanks a lot!